src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,399 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public static class ReadOnlyMemoryUtils
+    {
+
+        /// <summary>
+        /// This implements IEquatable's Equals method.
+        /// </summary>
+        public static bool Equals(ReadOnlyMemory<char> b, ReadOnlyMemory<char> memory)
+        {
+            if (memory.Length != b.Length)
+                return false;
+
+            Contracts.Assert(memory.IsEmpty == b.IsEmpty);
+
+            for (int i = 0; i < memory.Length; i++)
+            {
+                if (memory.Span[i] != b.Span[i])
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Compare equality with the given system string value.
+        /// </summary>
+        public static bool EqualsStr(string s, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValueOrNull(s);
+
+            if (s == null)
+                return memory.Length == 0;
+
+            if (s.Length != memory.Length)
+                return false;
+
+            for (int i = 0; i < memory.Length; i++)
+            {
+                if (s[i] != memory.Span[i])
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// For implementation of ReadOnlyMemory. Uses code point comparison.
+        /// Generally, this is not appropriate for sorting for presentation to a user.
+        /// </summary>
+        public static int CompareTo(ReadOnlyMemory<char> other, ReadOnlyMemory<char> memory)
+        {
+            int len = Math.Min(memory.Length, other.Length);
+            for (int ich = 0; ich < len; ich++)
+            {
+                char ch1 = memory.Span[ich];
+                char ch2 = other.Span[ich];
+                if (ch1 != ch2)
+                    return ch1 < ch2 ? -1 : +1;
+            }
+            if (len < other.Length)
+                return -1;
+            if (len < memory.Length)
+                return +1;
+            return 0;
+        }
+
+        public static IEnumerable<ReadOnlyMemory<char>> Split(char[] separators, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValueOrNull(separators);
+
+            if (memory.IsEmpty)
+                yield break;
+
+            if (separators == null || separators.Length == 0)
+            {
+                yield return memory;
+                yield break;
+            }
+
+            int ichMin = 0;
```

Comment:
>ichMin [](start = 16, length = 6)

As far as I see both `ichMin` and `ichLim` could, here and most methods in this file, be replaced with direct usage of the `0` literal and `memory.Length` (or, more precisely, `span.Length` since we should be storing the span upfront, see above) literal. This would both simplify the code and make it more clear.

Look at it this way. If we had code over an array `a` that looked like this:

```csharp
for (int i=0; i<a.Length; ++i)
    DoSomething(a[i]);
```

we certainly would not welcome any movement to change the code to this.

```csharp
int min = 0;
int lim = a.Length;
for (int i=min; i<lim; ++i)
    DoSomething(a[i]);
```

That's less clear and needlessly verbose.

Reply:
> ichLim might be worth keeping since we don't want to call memory.Length multiple times.

Why? Are you under the impression that calling this on a `ReadOnlySpan<char>` is a perf hit? Have you checked?

Just reactivating, since it reads like you aren't going to do this.

---
In reply to: [216189476](https://github.com/dotnet/machinelearning/pull/863#discussion_r216189476) [](ancestors = 216189476,216126883)

Reply:
No, I did not check but based on your comment for Span I assumed it might also be the case for Length but may be I'm wrong. I will replace ichLim as well.

---
In reply to: [216192013](https://github.com/dotnet/machinelearning/pull/863#discussion_r216192013) [](ancestors = 216192013,216189476,216126883)

Reply:
Span doing additional computation is also concerning to me. We need to take this up with the .NET team and provide this feedback. 

---
In reply to: [216194473](https://github.com/dotnet/machinelearning/pull/863#discussion_r216194473) [](ancestors = 216194473,216192140,216192013,216189476,216126883)